control-plane/connect-inject/peering_acceptor_controller.go

```diff
@@ -0,0 +1,259 @@
+// in Consul, it should be reconciled. If the status of the resource does not match the current state of the specified
+// secret, generate a new token and store it according to the spec.
+func (r *PeeringAcceptorController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("peering", req.NamespacedName)
+
+	r.Log.Info("received request for PeeringAcceptor:", "name", req.Name, "ns", req.Namespace)
+
+	// Get the PeeringAcceptor resource.
+	peeringAcceptor := &consulv1alpha1.PeeringAcceptor{}
+	err := r.Client.Get(ctx, req.NamespacedName, peeringAcceptor)
+
+	// If the PeeringAcceptor resource has been deleted (and we get an IsNotFound
+	// error), we need to delete it in Consul.
+	if k8serrors.IsNotFound(err) {
+		r.Log.Info("PeeringAcceptor was deleted, deleting from Consul", "name", req.Name, "ns", req.Namespace)
+		_, err := r.deletePeering(ctx, req.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to get PeeringAcceptor", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// Read the peering from Consul.
+	// Todo(peering) do we need to pass in partition?
+	readReq := api.PeeringReadRequest{Name: peeringAcceptor.Name}
+	peering, _, err := r.ConsulClient.Peerings().Read(ctx, readReq, nil)
+	var statusErr api.StatusError
+
+	// If the peering doesn't exist in Consul, generate a new token, and store it in the specified backend. Store the
+	// current state in the status.
+	if errors.As(err, &statusErr) && statusErr.Code == http.StatusNotFound && peering == nil {
+		r.Log.Info("peering doesn't exist in Consul", "name", peeringAcceptor.Name)
+
+		// Generate and store the peering token.
+		var resp *api.PeeringGenerateTokenResponse
+		if resp, err = r.generateToken(ctx, peeringAcceptor.Name); err != nil {
+			return ctrl.Result{}, err
+		}
+		if peeringAcceptor.Spec.Peer.Secret.Backend == "kubernetes" {
+			if err := r.createK8sPeeringTokenSecretWithOwner(ctx, peeringAcceptor, resp); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// Store the state in the status.
+		err := r.updateStatus(ctx, peeringAcceptor, resp)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		r.Log.Error(err, "failed to get Peering from Consul", "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// TODO(peering): Verify that the existing peering in Consul is an acceptor peer. If it is a dialing peer, an error should be thrown.
+	// If the peering does exist in Consul, compare the existing status to the spec, and decide whether to make updates.
+
+	// TODO(peering): If the status is empty, then what should be done? This means a peering exists in Consul but hasn't
+	// been reconciled by the controller, or there was an error during a previous reconciliation. I think this means we
+	// should continue and generate a new token and update the status with that.
```

Comment:
I like this idea and have added a ReconcileError to the status! In this case though, we want to know whether we have enough information in the status to get the previous secret value. I think even if we don't have a secret set in the status, we do want to generate a new token, because it means we've had an error in a previous reconcile or that the status update itself has failed. So I think if we do reach this case, we just do want to generate a new token. I've updated this behavior to do that

Reply:
that makes a lot of sense!